crates/ty_python_semantic/src/types.rs

```diff
@@ -3196,6 +3196,28 @@ impl<'db> Type<'db> {
                     .into()
                 };
 
+                let custom_getattribute_result = || {
+                    // Avoid cycles when looking up `__getattribute__`
+                    if "__getattribute__" == name.as_str() {
+                        return Symbol::Unbound.into();
+                    }
+
+                    // Typeshed has a `__getattribute__` method defined on `builtins.object` so we
+                    // explicitly hide it here using `MemberLookupPolicy::MRO_NO_OBJECT_FALLBACK`.
+                    self.try_call_dunder_with_policy(
+                        db,
+                        "__getattribute__",
+                        &mut CallArgumentTypes::positional([Type::StringLiteral(
+                            StringLiteralType::new(db, Box::from(name.as_str())),
+                        )]),
+                        MemberLookupPolicy::MRO_NO_OBJECT_FALLBACK,
```

Comment:
You also need to pass `MemberLookupPolicy::NO_INSTANCE_FALLBACK` here, because dunder methods should never be looked up on instances.

This is arguable a bad choice of API for `try_call_dunder_with_policy` (I will change that in a follow-up). `MemberLookupPolicy::NO_INSTANCE_FALLBACK` should always be implicitly added.

I also added a regression test that would have caught this.